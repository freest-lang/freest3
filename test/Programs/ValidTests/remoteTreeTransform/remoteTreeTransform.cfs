data Tree = Leaf | Node Int Tree Tree

transform :: forall α => Tree -> (rec x. +{LeafC: Skip, NodeC: !Int;x;x;?Int});α -> (Tree, α)
transform tree c =
  case tree of
    Leaf ->
      (Leaf, select LeafC c);
    Node x l r ->
      let c = select NodeC c in
      let c = send x c in
      let l1, c = transform[(rec x. +{LeafC: Skip, NodeC: !Int;x;x;?Int});?Int;α] l c in
      let r1, c = transform[?Int;α] r c in
      let x1, c = receive c in
      (Node x1 l1 r1, c)

treeSum :: forall α => (rec x. &{LeafC: Skip, NodeC: ?Int;x;x;!Int});α -> (Int, α)
treeSum c =
  match c with
    LeafC c -> (0, c);
    NodeC c ->
      let x, c = receive c in
      let l, c = treeSum[(rec x . &{LeafC: Skip, NodeC: ?Int;x;x;!Int});!Int;α] c in
      let r, c = treeSum[!Int;α] c in
      let c    = send (x+l+r) c in
      (x + l + r, c)

aTree :: Tree
aTree = Node 1 (Node 2 Leaf (Node 3 Leaf (Node 4 Leaf (Node 5 Leaf Leaf)))) (Node 6 Leaf (Node 7 Leaf (Node 8 Leaf Leaf)))

start :: Int
start =
  let w, r  = new (rec x. +{LeafC: Skip, NodeC: !Int;x;x;?Int}) in
  let t = fork (transform[Skip] aTree w) in
  let n, r1 = treeSum[Skip] r in
  n
  
{- This start requires a different .expect file:
   Node 7 Leaf (Node 4 Leaf Leaf)

start :: Tree
start =
  let aTree = Node 3 Leaf (Node 4 Leaf Leaf) in
  let w, r  = new (rec x . +{LeafC: Skip, NodeC: !Int;x;x;?Int}) in
  let x = fork (treeSum[Skip] r) in
    fst[Tree, Skip] (transform[Skip] aTree w)

-}
