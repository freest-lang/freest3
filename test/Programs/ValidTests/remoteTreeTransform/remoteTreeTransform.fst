data Tree = Leaf | Node Int Tree Tree

-- Note: we use the same constructors for the datatype and the channel, namely Leaf and Node

{-
  Writes a tree on a given channel;
  for each node in the tree reads an integer from the channel;
  returns a tree isomorphic to the input where each integer in nodes
  is read from the channel.
-}
transform : forall α => Tree -> (rec x. +{LeafC: Skip, NodeC: !Int;x;x;?Int});α -> (Tree, α)
transform tree c =
  case tree of
    Leaf ->
      (Leaf, select LeafC c);
    Node x l r ->
      let c = select NodeC c in
      let c = send x c in
      let l, c = transform[(rec x. +{LeafC: Skip, NodeC: !Int;x;x;?Int});?Int;α] l c in
      let r, c = transform[?Int;α] r c in
      let y, c = receive c in
      (Node y l r, c)

{-
  Reads a tree from a given channel;
  writes back on the channel the sum of the elements in the tree;
  returns this sum.
-}
treeSum : forall α => (rec x. &{LeafC: Skip, NodeC: ?Int;x;x;!Int});α -> (Int, α)
treeSum c =
  match c with
    LeafC c -> (0, c);
    NodeC c ->
      let x, c = receive c in
      let l, c = treeSum[(rec x. &{LeafC: Skip, NodeC: ?Int;x;x;!Int});!Int;α] c in
      let r, c = treeSum[!Int;α] c in
      let c    = send (x + l + r) c in
      (x + l + r, c)

aTree : Tree
aTree = Node 1 (Node 2 Leaf (Node 3 Leaf (Node 4 Leaf (Node 5 Leaf Leaf)))) (Node 6 Leaf (Node 7 Leaf (Node 8 Leaf Leaf)))

main : Tree
main =
  let w, r = new (rec x. +{LeafC: Skip, NodeC: !Int;x;x;?Int}) in
--  let t, w = fork (transform[Skip] aTree w) in
--  let n, r = treeSum[Skip] r in
  let _ = fork (treeSum[Skip] r) in
  let t, _ = transform[Skip] aTree w in
  t
