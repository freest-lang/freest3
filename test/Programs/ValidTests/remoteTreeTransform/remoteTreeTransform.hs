
data Tree = Leaf | Node Int Tree Tree
-- (rec xFormChan . +{LeafC: Skip, NodeC: !Int;xFormChan;xFormChan;?Int})



transform :: forall x => Tree -> (rec xFormChan . +{LeafC: Skip, NodeC: !Int;xFormChan;xFormChan;?Int});x -> (Tree, x)
transform tree c =
  case tree of
    Leaf ->
      (Leaf, select LeafC c)

    Node x l r ->
      let c1 = select NodeC c in
      let c2 = send x c1 in
      let l1, c3 = transform[rec xFormChan . +{LeafC: Skip, NodeC: !Int;xFormChan;xFormChan;?Int}] l c2 in
      let r1, c4 = transform[?Int;Skip] r c3 in
      -- let r1, c4 = transform[rec xFormChan . +{LeafC: Skip, NodeC: !Int;xFormChan;xFormChan;?Int}] r c3 in
      let x1, c5 = receive c4 in
      (Node x1 l1 r1, c5)



start :: Int
start = 10
         
      
  

