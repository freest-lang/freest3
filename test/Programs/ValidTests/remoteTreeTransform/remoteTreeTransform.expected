Node 7 Leaf (Node 4 Leaf Leaf)
